### ISSUE TITLE
expo-cookies library

### ISSUE THREAD

**Issue by brentvatne** [expo/expo#6756]
eg: https://github.com/expo/expo/issues/4380

Ideally code related to controlling cookies for WebKit (WKWebView) should live in react-native-webview. We should provide access to cookies that are included in `fetch` requests and other places where some Expo module interacts with http services. [react-native-cookies](https://github.com/react-native-community/react-native-cookies) is a good library to reference but "it would be better to use a per-project cookie storage instance (like a scoped unimodule would) and make sure all libraries that make network requests use that cookie storage instance. this requires providing expo-network and expo-image replacements for those respective libraries in RN".

**Comment by safaiyeh** [comment]
Let me know how I could help! I wouldn’t mind taking up some Expo work around cookies.

I’ve opened an issue in WebView asking the community about migrating the WebKit cookies there.

https://github.com/react-native-community/react-native-webview/issues/1139

**Comment by antoniogoulao** [comment]
Could this lib be a good candidate?
https://github.com/react-native-community/react-native-cookies

**Comment by brentvatne** [comment]
It would be helpful if this is a feature you would like to see if you could please provide information for what exactly you would use it to accomplish and what your current workaround for that functionality is. Thank you!

**Comment by brentvatne** [comment]
Main use cases from discussing with @safaiyeh:

- authenticate with react-native fetch with an api that only supports cookie based auth
- set cookie on webview based on the result of authenticating via fetch
- *in some cases* people use webview for auth and want to be able to extract cookies and use for react-native fetch. webview for auth is not allowed by apple but is fine on android.

More context:

> The biggest thing is react-native  fetch implementation does not manage cookies well. Cookies getting over written, not persisting the right ones, etc. My current use case is being able to persist cookies in AsyncStorage.

> Another thing, iOS session cookies do not automatically get put into requests. Have to manually get them from cookie storage and add them to requests

**Comment by casperstr** [comment]
Another use case that I am facing is when using expo-app-auth togheter with a oidc provider like auth0 that adds a cookie when signing in. If I want the possibility of another user signing in I need to delete the cookie before initialising another auth-process.

**Comment by NickCis** [comment]
I've found a workaround (only tested on Android via the Expo App) that allows the developer to manage fetch's cookies.

At least for the test that I've made (expo: `37.0.3`), if fetch is used with the option `credentials: 'omit'`, it respect the `cookie` header set in the `headers` property. So, if we use a cookie parsing + jar library (eg [tough cookie](https://www.npmjs.com/package/tough-cookie)), you can get a way to manage the cookies:

```js
import tough from 'tough-cookie';

export const cookiejar = new tough.CookieJar();

async function customFetch(url, opts = {}) {
  const cookies = await new Promise((rs, rj) => {
    cookiejar.getCookies(url, (err, cookies) => {
      if (err) {
        rj(err);
        return;
      }

      rs(cookies || []);
    });
  });

  const response = await fetch(url, {
    ...opts,
    headers: {
      ...opts.headers,
      cookies,
    },
  });

  if (response.headers['set-cookie']) {
    await new Promise((rs, rj) => {
      cookiejar.setCookie(
        response.headers['set-cookie'],
        request.url,
        (e, s) => {
          if (e) {
            rj(e);
            return;
          }

          rs(s);
        }
      );
    });
  }

  return response;
}

export default customFetch;

```

**Comment by desnor** [comment]
@NickCis I had a go at implementing something similar to your code above but wasn't able to use `tough-cookie` in my `expo` managed app just now because it relies on a native `node` module `url`, which is not present in the `react-native` javascript runtime. How did you get around that?

**Comment by NickCis** [comment]
@desnor React native has a limited implementation of Url, as far as I've checked it worked for using `tough-cookie`. You can check the code [here](https://github.com/NickCis/bypass-paywalls-app). Basically, in order to control all requests done in the web view, I'm inserting an script that overrides ajax (`fetch` and `XMLHttpRequest`) and listening for script creations and redirect everything to a message channel to end up doing requests with react-native's fetch (you can see the [webview code here](https://github.com/NickCis/bypass-paywalls-app/blob/master/src/ManagedWebView/index.tsx))

**Comment by desnor** [comment]
Thanks @NickCis that sounds pretty funky. I'm not using web views for my interactions, but instead the `AppAuth.authAsync` or related options (`WebBrowser.openAuthSessionAsync` or `AuthSession.startAsync`). It seems like the system browser that is opened during these sessions holds on to a cookie even after it is invalidated, and does not seek user credentials again for subsequent requests using these flows. So while it's handy to only have to sign in once, for a user who actively signs out, the expectation is that you would have to sign in again manually the next time. I'll have a look at your code though, might give me some ideas to think about.

**Comment by desnor** [comment]
> Another use case that I am facing is when using expo-app-auth togheter with a oidc provider like auth0 that adds a cookie when signing in. If I want the possibility of another user signing in I need to delete the cookie before initialising another auth-process.

@casperstr did you ever find any workaround for that behaviour? I'm encountering a similar issue, where after initially signing in via `AppAuth.authAsync` even after revoking the access token, subsequent sign in requests using the `AppAuth.authAsync` helper skip the user credentials interaction completely, and return a new access token with automatic redirection back into the app. So for a user who wanted to sign in with a different account, it is impossible.

**Comment by casperstr** [comment]
> > Another use case that I am facing is when using expo-app-auth togheter with a oidc provider like auth0 that adds a cookie when signing in. If I want the possibility of another user signing in I need to delete the cookie before initialising another auth-process.
> 
> @casperstr did you ever find any workaround for that behaviour? I'm encountering a similar issue, where after initially signing in via `AppAuth.authAsync` even after revoking the access token, subsequent sign in requests using the `AppAuth.authAsync` helper skip the user credentials interaction completely, and return a new access token with automatic redirection back into the app. So for a user who wanted to sign in with a different account, it is impossible.

I am using auth0, I discovered that they have a setting to disable SSO so that solved it for me.

**Comment by pdandradeb** [comment]
hey @brentvatne, I'd like to work on this issue. Is there any development already? Do you think is still possible to deliver on SDK 38?

In our case, users authenticate via SAML and cookies are set by many different hosts in the process and I'd like to be able to remove all previously set cookies when user logs out.

**Comment by brentvatne** [comment]
@pdandradeb - a workaround for now could be to have an endpoint to clear those cookies, and call that when signing out or before signing in. usually auth providers also give you some query param you can pass in to either force the user to auth again or to give them the option to, that could be another possibility here.

we won't have this ready for sdk 38, i don't think i have a clear picture of exactly what is needed here. if you'd like to help, a proposal that outlines the use cases that it intends to solve, the public api, and goes in detail on how the implementation would work would be the best way to get started

**Comment by pdandradeb** [comment]
Hey @brentvatne, thanks for your inputs. It seems indeed that Azure AD has a parameter to force reauthentication. I'll take a look on that.

I'm still considering a client side only solution, though.

Is already known which version of RN will be used in SDK 38? Networking was [exposed in RN 0.62.0]((https://github.com/react-native-community/releases/blob/master/CHANGELOG.md#v0620)) and I'm thinking that using `Networking.clearCookies()` will be the best option for me. 

I've tried to switch to incognito mode and back hoping it would get rid of the cookies but it looks like that would happen only in [android](https://github.com/react-native-community/react-native-webview/blob/b482bbd3a38d75e44830f68cda9053ef2a7958b6/android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java#L414), [not in iOS](https://github.com/react-native-community/react-native-webview/blob/b482bbd3a38d75e44830f68cda9053ef2a7958b6/apple/RNCWebView.m#L219).

**Comment by brentvatne** [comment]
@pdandradeb - sdk 38 will be using 0.62.2, so it looks like that method will be included:

https://github.com/expo/react-native/blob/28ec77da0aa7888228b752e18fb0926316094f9e/Libraries/Network/RCTNetworking.ios.js#L58-L59
https://github.com/expo/react-native/blob/28ec77da0aa7888228b752e18fb0926316094f9e/Libraries/Network/RCTNetworking.android.js#L86-L88

**Comment by pdandradeb** [comment]
Well, `RCTNetworking.clearCookies` only works in Android. The [iOS implementation](https://github.com/facebook/react-native/blob/d75401af9651193d83a2b89518e13ee5f73ee783/Libraries/Network/RCTNetworking.mm#L703) removes cookies from the app's shared cookie storage but [each WKWebView instance has its own cookie storage](https://developer.apple.com/documentation/foundation/nshttpcookiestorage?language=objc).

[Here's a recording](https://www.dropbox.com/s/gkbiriculzo8f1e/cookies-fail-with-networking-clearcookies.mov?dl=0) showing the behavior in both platforms.

I've created a PR https://github.com/react-native-community/react-native-webview/pull/1465 to include a `clearCookies` method in the `react-native-webview` on both platforms. Hope it will be merged.

Meanwhile, managed to configure Azure AD to enforce reauthentication using [ForceAuthn and IsPassive](https://docs.microsoft.com/en-us/azure/active-directory/develop/single-sign-on-saml-protocol) but still checking with our clients if we could change that configuration in production.

**Comment by Svarto** [comment]
@brentvatne currently there's no option (that I know of) to manage cookies in expo react-native. When a cookie is used for authentication (e.g. refresh token is stored in a cookie) which works great for web, in react-native mobile I don't have ability to control if the cookie is stored, how long, to delete it or to include it or not in certain requests. I think that's what is being asked for, ability to control cookies similar to react-native-cookies library.

For example, I get authentication with user and password through endpoint /auth/login with a normal axios / fetch and with the response comes the JWT token in json format and the refresh token comes as a cookie. To then refresh the JWT token I just fetch endpoint /auth/refresh with a request only including the cookie and the refresh token is then matched server side to to validate that my JWT tokenbrent can be refreshed.

Does this help when it comes to use case brentvatne?

**Comment by hbendix** [comment]
This would be greatly appreciated. I have an issue, [detailed here](https://github.com/expo/expo/issues/9213), which is preventing us from using iOS.

**Comment by amsimoes** [comment]
@NickCis using that fetch and tough cookie gives access to HttpOnly cookies?

**Comment by amsimoes** [comment]
Is it out yet? Or any ETA for which sdk it will be included?
Will it be able to fix this [issue](https://github.com/react-native-cookies/cookies/issues/76) on Android?

**Comment by brentvatne** [comment]
we haven't had time to work on this. we recommend not depending on cookies for authentication in your app right now and using token-based auth instead.

**Comment by MartinMasek** [comment]
@brentvatne We use token based auth but the issue is when a user logs out and a different user wants to login. Because we are using WebBrowser.openAuthSessionAsync, the browser keeps the session in the browser and no password is needed. We used @react-native-community/cookies to deal with this but haven't figure out how to do it in managed Expo flow.

**Comment by summerkiflain** [comment]
I have another use-case i guess, I have an expo app that syncs user's data using multiple parallel axios calls to my `Node.js` back-end server which is hosted on Heroku. (I am using multiple axios calls to show user progress on syncing process).

**IDEA:** I am thinking of switching this process to sockets, because as user's data increase its take more time to sync using multiple parallel calls, and it blocks some of the more urgent axios calls which user want to see first, a detail page of some listing item is probably more important to user to see than the lots of data syncing first before showing detail page.

**Options:** I have probably two options either add a priority system for axios calls, and then set lower priority for syncing calls and higher priority for all other calls (couldn't find any good lib for this too), or switch syncing data to sockets so that axios calls are not blocked by the syncing process.

**Problem:** Issue is my back-end Node.js server is hosted on Heroku and Heroku uses [dynos](https://www.heroku.com/dynos), when enabling sockets support on heroku, its has two options, ([Option 1: WebSocket](https://devcenter.heroku.com/articles/node-websockets#option-1-websocket) OR [Option 2: Socket.io](https://devcenter.heroku.com/articles/node-websockets#option-2-socket-io)) 

its says to do this:
```
heroku features:enable http-session-affinity
```

**Comment by summerkiflain** [comment]
I have another use-case i guess, I have an expo app that syncs user's data using multiple parallel axios calls to my `Node.js` back-end server which is hosted on Heroku. (I am using multiple axios calls to show user progress on syncing process).

**IDEA:** I am thinking of switching this process to sockets, because as user's data increase its take more time to sync using multiple parallel calls, and it blocks some of the more urgent axios calls which user want to see first, a detail page of some listing item is probably more important to user than the lots of data syncing first before showing detail page.

**Options:** I have probably two options either add a priority system for axios calls, and then set lower priority for syncing calls and higher priority for all other calls (couldn't find any good lib for this too), or switch syncing data to sockets so that axios calls are not blocked by the syncing process.

**Problem:** Issue is my back-end `Node.js` server is hosted on Heroku and Heroku uses [dynos](https://www.heroku.com/dynos), when enabling sockets support on heroku, its has two options, ([Option 1: WebSocket](https://devcenter.heroku.com/articles/node-websockets#option-1-websocket) OR [Option 2: Socket.io](https://devcenter.heroku.com/articles/node-websockets#option-2-socket-io)) 
I guess I can only use `socket.io` option for the syncing process not sure about the ws protocol support on expo, guide me if i am wrong.

using `Socket.io`, Heroku doc's says to do this:

>Apps using Socket.io should enable [session affinity](https://devcenter.heroku.com/articles/session-affinity). If you plan to use Node’s Cluster module or to scale your app to multiple dynos, you should also follow Socket.io’s [multiple-nodes](http://socket.io/docs/using-multiple-nodes/) instructions.

```
heroku features:enable http-session-affinity
```
and in [How it works](https://devcenter.heroku.com/articles/session-affinity#how-it-works) of session-affinity this is what I am concerned for:
>It works on arbitrary clients behind arbitrary ISPs or infrastructure, as long as they support HTTP cookies.

>As soon as the feature is enabled on an application, the Heroku router will start adding an HTTP cookie named heroku-session-affinity to every new request and response.

Please read [How it works](https://devcenter.heroku.com/articles/session-affinity#how-it-works) of session-affinity for more info.

I am just concerned, if cookies are not supported yet by expo, using `socket.io` with Heroku platform is also not going to work. I'd like some insight on this, please let me know if someone is using scokets reliably in expo + Heroku setup. Regards.

**Comment by antonmamonov** [comment]
I have a use-case as well, I'm using an expo app that requires access to all the cookies stored in the app's webview. As a workaround, we use the `@react-native-community/cookies` library in our ejected expo app using the `CookieManager.getAll` function. 

We would like to switch over to the non-ejected version, given that would greatly simplify the entire workflow of our app.

@brentvatne is there any ETA on having this incorporated in a future version of Expo?

**Comment by Dakuan** [comment]
Lack of cookie clearance is making it almost impossible to develop an expo app with Auth0. The only way of forcing a new login is by also forcing a new signup by deleting the user in Auth0.

**Comment by AkwasiSB** [comment]
@Dakuan thanks very much

**Comment by SampsonCrowley** [comment]
@brentvatne is there some way to at least get `@react-native-community/cookies` integrated without ejecting until you have time to work on it here? 

Token based auth without being able to use `httpOnly` cookies to store a device identifier opens up a whirlwind of possibilities for using stolen JWTs and basically nullifies the benefit of having them for web clients; since then a hacker could just use the app endpoint to login instead, because the app endpoint would need to disable setting a device identifier in `httpOnly`

**Comment by brentvatne** [comment]
we can't just install `@react-native-community/cookies` because this would be a significant security concern within expo go, given that you can load multiple different projects (including those hosted on snack) within the app. so sandboxing is necessary, similar to what we have to do with apis like asyncstorage and all permissions.

sorry, i hope you can find a workaround that works for you for now. this isn't in progress at the moment - we understand that something like this is important for us to provide a solid solution for in the future, but it's a big project and we're focusing on solving the underlying generalized problem of letting people include custom native code in their managed apps. we're hoping to ship a preview of the features that will make it possible for you to include libraries like `@react-native-community/cookies` in your own custom client build around q3 2021.

**Comment by antonmamonov** [comment]
@brentvatne thank you so much for getting back to me and the additional info! We're currently bottlenecked with the ejected expo so having `@react-native-community/cookies` working with the standard un-ejected version would really help us out given we don't have a workaround. We'll keep an eye out for your solution coming out in Q3. If you end shipping a solution before Q3 please let us know. Appreciate it!

**Comment by arcticmatt** [comment]
My use case is that I want to be able to reuse the authentication logic I already have, which currently supports cookies.

Specifically, I want to be able to do something like https://build.affinity.co/persisting-sessions-with-react-native-4c46af3bfd83. E.g.

```
// Store cookie in SecureStore
const response = await fetch('login_endpoint', {...});
const cookies = setCookie.parse(response.headers.get("set-cookie"));
await SecureStore.setItemAsync("cookie", cookies[0]);

// Use cookie for subsequent requests
const cookie = await SecureStore.getItemAsync("cookie");
const authenticatedResponse = await fetch('authenticated_endpoint', {
  headers: {
    cookie: `${cookie.name}=${cookie.value}`,
  }
});
```

Based on reading https://build.affinity.co/persisting-sessions-with-react-native-4c46af3bfd83 and https://locastic.com/blog/react-native-cookie-based-authentication/, it seems like I should use react-native-cookies to clear cookies stored natively before making the second request, e.g. something like

```
import CookieManager from 'react-native-cookies'
await CookieManager.clearAll();

const cookie = await SecureStore.getItemAsync("cookie");
const authenticatedResponse = await fetch('authenticated_endpoint', {
  headers: {
    cookie: `${cookie.name}=${cookie.value}`,
  }
});
```


@pdandradeb now that your PR has landed, do you know if calling `RCTNetworking.clearCookies` is a sufficient replacement for calling `CookieManager.clearAll`?

Also @brentvatne do you know of any way I can do something like this without having to eject and use react-native-cookies?

**Comment by brentvatne** [comment]
there isn't a good way to accomplish that right now. while not ideal, the easiest option may be to add token based auth to your backend

**Comment by arcticmatt** [comment]
Gotcha, thanks for the reply @brentvatne. Do you know why calling `RCTNetworking.clearCookies` is not sufficient?

FWIW I've implemented the code I wrote above and it seems to be working. Although based on the blog posts, it seems like it might break in certain situations.

**Comment by brentvatne** [comment]
you may have issues with cookies persisting between app loads, it's been reported here https://github.com/expo/expo/issues/9213

**Comment by arcticmatt** [comment]
For #9213, I don't see any mention of clearing the cookies with `RCTNetworking.clearCookies` before each request. So seems like that issue doesn't tell us much about the effectiveness of `RCTNetworking.clearCookies` either way

**Comment by brentvatne** [comment]
@arcticmatt - i have no comment on the effectiveness of `RCTNetworking.clearCookies` - rather i was just saying that it is a moot point if you can't persist cookies reliably between app loads

**Comment by giautm** [comment]
Hello, I want to clear all Cookie in WebView.

Because My app has embed websites of third-party and they use Phone number OTP to authentication, we can't make any change on whose side. So, I have to clear all related cookies if exist when my user log out of the app. I found a [comment on StackOverflow](https://stackoverflow.com/a/61825519/5853096) that says the bellow method maybe work. And [another link](https://github.com/react-native-cookies/cookies#platforms-supported) direct me to here, I have confused. 😕 

Is this method will work?
```js
const RCTNetworking = require('react-native/Libraries/Network/RCTNetworking')
RCTNetworking.clearCookies(() => { })
```

**Comment by Stevemoretz** [comment]
So any idea about the progress of this module?

**Comment by robwatkin** [comment]
Hi all, I am new to ReactNative and Expo. I am evaluating Expo to build an app against an existing API which uses cookie based authentication. We have a prototype working which can login, and access API endpoints but it only works on IOS and Android but not Web (currently Chrome on MacOS). It fails because `fetch` does not save the cookie on login and present it on future api calls. Is my understanding (from this thread) that this is currently not possible (without nasty hacks and/or eject the project) correct? Thanks, Rob :)

**Comment by PedroBern** [comment]
+1

**Comment by awgrass** [comment]
+1

**Comment by JonathanEkberg** [comment]
+1

**Comment by lottemarines** [comment]
+1

**Comment by binajmen** [comment]
Hi,

After reading the whole thread, could someone confirm my understanding: it is **not** possible to use a cookie-based API at the moment using Expo?

How people are handling that? Is this a requirement to eject?

**Comment by brentvatne** [comment]
@binajmen - you can use any api you need with eas build / development builds: https://docs.expo.dev/development/introduction/

cookies are not part of the expo sdk but you're not limited to libraries that we build, that's just a starting point and what we ship in expo go

**Comment by binajmen** [comment]
Hi @brentvatne, thank you for taking the time to answer my silly question. However, I'm not sure to understand your suggestions 😕 (I'm new in this react native world, expo is awesome but I'm a bit lost regarding the cookie topic)

**Comment by Codelica** [comment]
@binajmen  depending on how you're going to use them, cookies do "work" in React Native / Expo.  However, there are some quirks and it's important to learn what the boundaries are (WebView + fetch vs WebBrowser).  Our application uses uses cookies and OAuth for authentication and sessions across several separate backends.  That said, I've run into some issues requiring workarounds.  For example we use Websockets to implement a streaming API.  On iOS cookies are provided as expected on Websocket  calls regardless of path.  On Android only cookies that are provided on the root path (/) are sent regardless of requested path.  I've also run into an issue with timing when a cookie is initially set.  I'm sure there are other things, but in general they are working well (enough) for us.

**Comment by binajmen** [comment]
Hi @Codelica, thank you for sharking your experience.

Long story short, I need to connect to a backend that requires the typical email/password in exchange for a cookie that should be used in the header (`Set-Cookie`) on every restricted API call.

Perhaps I was confused by everything I've read here and there about the cookies being overwritten by the native "engine", leading to unstable situations?

I suppose the best would be to try and conquer (or fail 😄).

**Comment by Codelica** [comment]
@binajmen that's generally how I roll.  ;)  But in general you should be fine making a login form (in RN/Expo) and then doing a fetch call to log them in and have a cookie set.  That cookie will then be provided on appropriate fetch calls or WebViews where you set `sharedCookiesEnabled={true}` (not to be confused with WebBrowser that uses a different cookie store).

**Comment by binajmen** [comment]
@Codelica what if it is REST API endpoints I need to hit to login and/or retrieve data? Is WebView still the way to go? 🧐

**Comment by Codelica** [comment]
@binajmen you'll be hitting all your REST endpoints with fetch() calls.  If a cookie is set on a fetch call it will be available on subsequent fetch calls with `credentials: 'include'` set, and be available to related URLs displayed in a WebView with `sharedCookiesEnabled={true}` set.  Once you give it a try, you'll see.   Don't want to hijack this issue with 27 subs on it though.  The [Expo forums](https://forums.expo.dev/) helped me a lot when I was getting started.  Good luck!

**Comment by brentvatne** [comment]
we don't plan on working on this any time soon, we recommend using @react-native-cookies/cookies :) https://github.com/react-native-cookies/cookies#expo

**Comment by Stevemoretz** [comment]
> we don't plan on working on this any time soon, we recommend using @react-native-cookies/cookies :) https://github.com/react-native-cookies/cookies#expo

It's funny you and Evan both have contributed there just saw it. Sort of an expo project if you two have contributed anyways lol.

Thank you so much for just putting an end to this God knows how many times I've come here to check what's up.
